client/jmx-adapter/src/main/java/org/jolokia/client/jmxadapter/ToOpenTypeConverter.java

```diff
@@ -0,0 +1,439 @@
+   * coded tabular return types (important for visual presentation in certain tools) 4. Use cached
+   * type for attribute/item : either hardcoded to please JConsole / JVisualVM or introspected from
+   * MBeanInfo 5. Dynamically build structured type from contents (will struggle with null values
+   * for unknown entities) 6. Fail
+   */
+  public static OpenType<?> recursivelyBuildOpenType(String name, Object rawValue)
+      throws OpenDataException {
+    for (SimpleType<?> type : typeArray) {
+      if (type.isValue(rawValue)
+          || (type.getClassName() != null && type.equals(cachedType(name)))) {
+        return type;
+      }
+    }
+    if (rawValue instanceof JSONArray) {
+      final JSONArray array = (JSONArray) rawValue;
+      if (array.size() > 0) {
+        final OpenType<?> elementType = recursivelyBuildOpenType(name + ".item", array.get(0));
+        if (elementType instanceof SimpleType && cachedType(name) != null) {
+          return cachedType(name);
+        } else {
+          return ArrayType.getArrayType(elementType);
+        }
+      }
+    } else if (tabularContentType(name) != null) {
+      final String typeName =
+          "Map<java.lang.String," + tabularContentType(name).getClassName() + ">";
+      return new TabularType(
+          typeName,
+          typeName,
+          new CompositeType(
+              typeName,
+              typeName,
+              new String[]{"key", "value"},
+              new String[]{"key", "value"},
+              new OpenType<?>[]{STRING, tabularContentType(name)}),
+          new String[]{"key"});
+
+    } else if (cachedType(name) != null) {
+      return cachedType(name);
+    } else if (rawValue instanceof JSONObject) {
+      final JSONObject structure = (JSONObject) rawValue;
+      final String[] keys = new String[structure.size()];
+      final OpenType<?>[] types = new OpenType[structure.size()];
+      int index = 0;
+      for (Object element : structure.entrySet()) {
+        @SuppressWarnings("unchecked")
+        Map.Entry<String, Object> entry = (Entry<String, Object>) element;
+        keys[index] = entry.getKey();
+        types[index++] = recursivelyBuildOpenType(name + "." + entry.getKey(), entry.getValue());
+      }
+      if (types.length == 0) {
+        throw new InvalidOpenTypeException("No subtypes for " + name);
+      } else {
+        return new CompositeType("complex", "complex", keys, keys, types);
+      }
+    }
+    // should probably never happen, to signify type could not be found
+    throw new InvalidOpenTypeException("Unable to figure out type for " + rawValue);
+  }
+
+  static OpenType<?> cachedType(final String name) throws OpenDataException {
+    if (TYPE_SPECIFICATIONS == null) {
+      TYPE_SPECIFICATIONS = new HashMap<String, OpenType<?>>();
+      //Specifically override types of some central Java types to suit JConsole and jvisualvm tools
+      //overrides follow QName.attribute.innerAttribute recursively
+      cacheType(
```

Comment:
'guess this will needs some constant maintain efforts for new JDK to come, but I'm fine with it as I don't know how to do it better :)

Reply:
Well, we could use TestNG groups or something, if they should be handled different than ci tests. Also as long as we require 1.6 compatibility we can no longer build with the newest JDK releases, so I have mostly relied on manual testing with jmc/visualvm/jconsole for that.

Reply:
At the same time it is good to test everything available in the JMX tree somehow.

Reply:
I have now run the tests on all Oracle and openjdk releases 7-14, and cover all of those